### PR TITLE
Add Reshape Op

### PR DESCRIPTION
**Pull request by klecki**
Implementation based on Copy operator.
Changes the shape of output tensor
and copies the data to output.

TODO:
* testing
* Handling of shapes on GPU memory?

Futures:
* Allow for Arguments to have more dimensions ?
* Better handling of vectors as arguments for all the types
that can appear in DALI ?

Signed-off-by: Krzysztof Lecki <klecki@nvidia.com>

### CODE REVIEW

File: dali/pipeline/operators/util/reshape.cu [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  auto &input = ws->Input<GPUBackend>(kInTensorIdx);
+  auto *output = ws->Output<GPUBackend>(kOutTensorIdx);
+  auto new_shapes = GetNewShapesForSamples(ws);
+  output->set_type(input.type());
+  output->Resize(new_shapes);
+  for (size_t i = 0; i < new_shapes.size(); i++) {
+    DALI_ENFORCE(Product(new_shapes[i]) == Product(input.tensor_shape(i)),
+                 "New shape must represent exactly the same number of elements as old shape.");
+  }
+  CUDA_CALL(cudaMemcpyAsync(output->raw_mutable_data(), input.raw_data(), input.nbytes(),
```
Comment (Kh4L) [inline dali/pipeline/operators/util/reshape.cu]: Why not using `type.Copy<GPUBackend, GPUBackend>` also here?

File: dali/pipeline/operators/util/reshape.cc [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  auto &input = ws->Input<CPUBackend>(kInTensorIdx);
+  auto *output = ws->Output<CPUBackend>(kOutTensorIdx);
+  std::vector<Index> new_sample_shape = GetNewShapeForSample(ws);
+  DALI_ENFORCE(Product(new_sample_shape) == Product(input.shape()),
+               "New shape must represent exactly the same number of elements as old shape.");
+  output->set_type(input.type());
+  output->Resize(new_sample_shape);
+
+  TypeInfo type = input.type();
+  type.Copy<CPUBackend, CPUBackend>(output->raw_mutable_data(), input.raw_data(), input.size(), 0);
```
Comment (JanuszL) [inline dali/pipeline/operators/util/reshape.cc]: Why copy and not `output->ShareData`

Comment (klecki) [reply dali/pipeline/operators/util/reshape.cc]: As we discussed, it is not used and may interfere with how Executor works on the boundaries of stages. 
Reverting to Copy.

File: dali/pipeline/operators/util/reshape.cu [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+namespace dali {
+
+template <>
+std::vector<std::vector<Index>> Reshape<GPUBackend>::GetNewShapesForSamples(DeviceWorkspace *ws) {
+  auto num_samples = ws->Input<CPUBackend>(kInTensorIdx).ntensor();
+  std::vector<std::vector<Index>> result;
+  if (use_input_for_shapes_) {
+    // TODO(klecki) is there a way that we could handle GPU shapes here?
+    auto &new_shapes = ws->Input<CPUBackend>(kInShapeIdx);
```
Comment (Kh4L) [inline dali/pipeline/operators/util/reshape.cu]: const

File: dali/pipeline/operators/util/reshape.cu [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    DALI_ENFORCE(new_shapes.IsDenseTensor() && new_shapes.tensor_shape(0).size() == 1,
+                 "New shape should be represented as 1-dim tensor with uniform dimension");
+    int dim = new_shapes.tensor_shape(0)[0];
+    for (decltype(num_samples) i = 0; i < num_samples; i++) {
```
Comment (Kh4L) [inline dali/pipeline/operators/util/reshape.cu]: Why auto and decltype here? You don't trust ntensor?

File: dali/pipeline/operators/util/reshape.cu [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+      result.push_back(new_shape_);
+    }
+  }
+  return result;
+}
+
+template <>
+void Reshape<GPUBackend>::RunImpl(DeviceWorkspace *ws, const int idx) {
+  auto &input = ws->Input<GPUBackend>(kInTensorIdx);
```
Comment (Kh4L) [inline dali/pipeline/operators/util/reshape.cu]: const

File: dali/pipeline/operators/util/reshape.cu [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+  auto *output = ws->Output<GPUBackend>(kOutTensorIdx);
```
Comment (Kh4L) [inline dali/pipeline/operators/util/reshape.cu]: Rebase onto master to get the Workspace::Output lvalue ref update

File: dali/pipeline/operators/util/reshape.cu (hunk 1 above)
Comment (klecki) [reply dali/pipeline/operators/util/reshape.cu]: Done

File: dali/pipeline/operators/util/reshape.cu (hunk 3 above)
Comment (klecki) [reply dali/pipeline/operators/util/reshape.cu]: Done.

File: dali/pipeline/operators/util/reshape.cu (hunk 5 above)
Comment (klecki) [reply dali/pipeline/operators/util/reshape.cu]: Done

File: dali/pipeline/operators/util/reshape.cu (hunk 4 above)
Comment (klecki) [reply dali/pipeline/operators/util/reshape.cu]: Yeah, the problem is, we sometimes have signed sizes and sometimes unsigned.

**Comment by klecki** [comment]
Build: 620404

**Comment by klecki** [comment]
Build: 623307

**Comment by klecki** [comment]
Build: 623329

File: dali/pipeline/operators/util/reshape.cc [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  type.Copy<CPUBackend, CPUBackend>(output.raw_mutable_data(), input.raw_data(), input.size(), 0);
+}
+
+DALI_REGISTER_OPERATOR(Reshape, Reshape<CPUBackend>, CPU);
+
+DALI_SCHEMA(Reshape)
+    .DocStr("Reshape the tensor")
+    .NumInput(1, 2)
+    .NumOutput(1)
+    .AddOptionalArg("new_shape", R"code(New shape for each sample.
```
Comment (JanuszL) [inline dali/pipeline/operators/util/reshape.cc]: It seems we have new requirements - per sample reshape. So imaging that we have TF record with raw data in one entry, and shape in the other one, then we need to read both entries and perform per sample reshape.

Comment (JanuszL) [reply dali/pipeline/operators/util/reshape.cc]: And we need probably support two cases:
- shape is given as 2D tensor
- shape is given as 2, 1D tensors

**Comment by klecki** [comment]
Scope change, need to be reworked.

**Pull request closed without merging**